### ISSUE TITLE
Classification in C++ with ONNX or OpenCV

### ISSUE THREAD

**Issue by sctrueew** [ultralytics/yolov5#10737]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi,
Is it possible to inference classify in ONNX C++ or OpenCV Dnn?

**Comment by VITEK-THE-BEST** [comment]
Hello friend! I also needed to solve this problem. Got the following solution:

```python
from sklearn.utils.extmath import softmax

net = cv2.dnn.readNetFromONNX("best.onnx")

frame = cv2.imread(f'img.png')
frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB) 
frame = cv2.resize(frame,(640,640)) #IMAGE SIZE

MEAN = 255 * np.array([0.485, 0.456, 0.406])
STD = 255 * np.array([0.229, 0.224, 0.225])
x = np.array(frame)
x = x.transpose(-1, 0, 1)
x = (x - MEAN[:, None, None]) / STD[:, None, None]

net.setInput(x.reshape(1,3,640,640))

preds = net.forward()
```

~~But I'm not a particularly good programmer, and I see that the accuracy is lower than what it gives out `classify/predict.py `. If you know the option better, I will be very glad to know your version!~~

[It helped me](https://stackoverflow.com/a/69630766/16869802)

***
### Update
the code is fully working, the file `classify/predict.py ` returns a rounded value!

**Comment by sctrueew** [comment]
@VITEK-THE-BEST Hi, I have implemented it in C++ but the result isn't right.


      `void loadObj()
       {
	std::cout << "OCV model loading..." << std::endl;

	net = cv::dnn::readNetFromONNX(config.modelPath);

	if (config.isGPU) {
		net.setPreferableBackend(cv::dnn::DNN_BACKEND_CUDA);
		net.setPreferableTarget(cv::dnn::DNN_TARGET_CUDA_FP16);
	}
	//cpu
	else {
		net.setPreferableBackend(cv::dnn::DNN_BACKEND_DEFAULT);
		net.setPreferableTarget(cv::dnn::DNN_TARGET_CPU);
	}

	std::cout << "OCV model loaded" << std::endl;
     }


      template <typename T>
      std::vector<size_t> SortIndexes(const std::vector<T>& v) {
      
	      // initialize original index locations
	      std::vector<size_t> idx(v.size());
	      std::iota(idx.begin(), idx.end(), 0);
      
	      // sort indexes based on comparing values in v
	      std::sort(idx.begin(), idx.end(),
		      [&v](size_t i1, size_t i2) {return v[i1] > v[i2]; });
      
	      return idx;
      }


    std::vector<MlResult> Detect(cv::Mat frame)
    {
	    std::vector<MlResult> result;
    
	    cv::Scalar mean{ 0.485, 0.456, 0.406 };
	    cv::Scalar std{ 0.229, 0.224, 0.225 };
	    bool swapRB = false;
	    bool crop = false;
	    int inpWidth = 224;
	    int inpHeight = 224;
	    
	    cv::Mat blob;
	    cv::dnn::blobFromImage(frame, blob, 1.0, cv::Size(inpWidth, inpHeight), mean, swapRB, crop);
	    if (std.val[0] != 0.0 && std.val[1] != 0.0 && std.val[2] != 0.0) {
		    cv::divide(blob, std, blob);
	    }
    
	    net.setInput(blob);
	    cv::Mat prob = net.forward();
	    
	    cv::Mat probReshaped = prob.reshape(1, prob.total() * prob.channels());
	    std::vector<float> probVec =
		    probReshaped.isContinuous() ? probReshaped : probReshaped.clone();
		    
	    std::vector<float>vec(probReshaped.begin<float>(), probReshaped.end<float>());
	    auto sorted = SortIndexes(vec);
    
	    //auto res = ox_softmax(indexValuePairs, numClasses);
	    const int topk = std::min(config.topK, (int)sorted.size());
    
	    for (size_t i = 0; i < topk; ++i) {
		    const auto& resultd = sorted[i];
		    const auto& conf = vec[resultd];
		    MlResult mr;
		    mr.score = conf / 100.0;
		    mr.class_id = resultd;
		    result.push_back(mr);
	    }
    
    
	    return result;
    }
    
    
    void dispose() {
	    // Destroy the engine
	    net.~Net();
    }`

**Comment by VITEK-THE-BEST** [comment]
@sctrueew 
I rechecked my python code, it works correctly, the file `classify/predict.py ` rounds the values. We can say that the python code works correctly 

I'm not good at C++. But tomorrow I will try to rewrite my code in GO. As I rewrite, I will write here

**Comment by VITEK-THE-BEST** [comment]
```go
package main

import (
	"fmt"
	"image"
	"math"
	"sync"

	"gocv.io/x/gocv"
)

func main() {

	net := NewYoloModelCls("best.onnx")

	frame := gocv.IMRead("3_resize.png", gocv.IMReadColor)
	defer frame.Close()

	gocv.Resize(frame, &frame, image.Point{640, 640}, 0, 0, gocv.InterpolationLinear)

	result, err := net.Invoke(&frame)
	if err != nil {
		panic(err)
	}
	defer net.Close()

	fmt.Println(result)

}

type YoloModelCls struct {
	net   *gocv.Net
	mutex sync.Mutex
}

func NewYoloModelCls(model_path string) YoloModelCls {
	net := gocv.ReadNetFromONNX(model_path)

	net.SetPreferableBackend(gocv.NetBackendOpenCV)
	net.SetPreferableTarget(gocv.NetTargetCPU)

	return YoloModelCls{
		net: &net,
	}
}

func (model *YoloModelCls) Invoke(frame *gocv.Mat) (float64, error) {
	model.mutex.Lock()
	defer model.mutex.Unlock()

	blob := gocv.BlobFromImage(*frame, 1, image.Pt(frame.Rows(), frame.Cols()), gocv.NewScalar(0, 0, 0, 0), true, false)
	defer blob.Close()

	err := normalize(&blob)
	if err != nil {
		return 0.0, err
	}

	model.net.SetInput(blob, "")

	preds := model.net.Forward("")
	defer preds.Close()

	normal_preds, err := preds.DataPtrFloat32()
	if err != nil {
		return 0.0, err
	}
	result := softmax(normal_preds)

	return result[0], nil
}

func (model *YoloModelCls) Close() {
	model.net.Close()
}

func normalize(blob *gocv.Mat) error {

	arr_matrix, err := blob.DataPtrFloat32()
	if err != nil {
		return err
	}

	x := blob.Size()[2]
	y := blob.Size()[3]

	mean := [3]float32{123.675, 116.28, 103.53}
	std := [3]float32{58.395, 57.12, 57.375}

	for i := 0; i < x*y; i++ {
		arr_matrix[i] = (arr_matrix[i] - mean[0]) / std[0]
	}

	for i := x * y; i < x*y*2; i++ {
		arr_matrix[i] = (arr_matrix[i] - mean[1]) / std[1]
	}

	for i := x * y * 2; i < x*y*3; i++ {
		arr_matrix[i] = (arr_matrix[i] - mean[2]) / std[2]
	}

	return nil
}

//softmax by 2 classes
func softmax(preds []float32) [2]float64 {
	exp_preds := [2]float64{}
	exp_preds[0] = math.Exp(float64(preds[0]))
	exp_preds[1] = math.Exp(float64(preds[1]))

	sum := exp_preds[0] + exp_preds[1]

	exp_preds[0] /= sum
	exp_preds[1] /= sum

	return exp_preds
}
```

**Comment by sctrueew** [comment]
@VITEK-THE-BEST Thanks. I'll try it
